Fix #84 - Revise development documentation (#158)
nautobot/nautobot: nautobot/docs/plugins/development.md

### Create setup.py

> The`setup.py` script is the [setup script](https://docs.python.org/3.6/distutils/setupscript.html) we'll use to install our plugin once it's finished. The primary function of this script is to call the setuptools library's `setup()` function to create a Python distribution package. We can pass a number of keyword arguments to inform the package creation as well as to provide metadata about the plugin. 

Reviewer: I imagine we'll want to revise this section at some point to use Poetry and `pyproject.toml` instead of `setup.py`, since all of our "official" plugins use those.

Author: Yes, agreed. I just left it alone for now until we're ready to do that. It's on my list.

Author: I'm going to do the plugin bootstrap docs for setup.py -> pyproject.toml in another PR.

config = AnimalSoundsConfig
```

Nautobot looks for the `config` variable within a plugin's `__init__.py` to load its configuration. Typically, this will be set to the `PluginConfig` subclass, but you may wish to dynamically generate a `PluginConfig` class based on environment variables or other factors.

#### PluginConfig Attributes

The configurable attributes for a `PluginConfig` are listed below in alphabetical order.

| Name | Description |
| ---- | ----------- |
| `author` | Name of plugin's author |
| `author_email` | Author's public email address |
| `base_url` | (Optional) Base path to use for plugin URLs. If not specified, the project's `name` will be used. |
| `caching_config` | Plugin-specific cache configuration |
| `custom_validators` | The dotted path to the list of custom validator classes (default: `custom_validators.custom_validators`) |
| `datasource_contents` | The dotted path to the list of datasource (Git, etc.) content types to register (default: `datasources.datasource_contents`) |

> | `description` | Brief description of the plugin's purpose |
> | `graphql_types` | The dotted path to the list of GraphQL type classes (default: `graphql.graphql_types)` |
> | `default_settings` | A dictionary of configuration parameters and their default values |

Reviewer: ```suggestion
| `default_settings` | A dictionary of configuration parameters and their default values |
| `description` | Brief description of the plugin's purpose |
| `graphql_types` | The dotted path to the list of GraphQL type classes (default: `graphql.graphql_types)` |
```

Author: Why am I so bad at this? It's to keep you busy, that's why!

| `installed_apps` | A list of additional Django application dependencies to automatically enable when the plugin is activated (you must still make sure these underlying dependent libraries are installed) |
| `jobs` | The dotted path to the list of Job classes (default: `jobs.jobs`) |

> | `menu_items` | The dotted path to the list of menu items provided by the plugin (default: `navigation.menu_items`) |
> | `max_version` | Maximum version of Nautobot with which the plugin is compatible |
> | `min_version` | Minimum version of Nautobot with which the plugin is compatible |
> | `middleware` | A list of middleware classes to append after Nautobot's built-in middleware |

Reviewer: ```suggestion
| `max_version` | Maximum version of Nautobot with which the plugin is compatible |
| `menu_items` | The dotted path to the list of menu items provided by the plugin (default: `navigation.menu_items`) |
| `middleware` | A list of middleware classes to append after Nautobot's built-in middleware |
| `min_version` | Minimum version of Nautobot with which the plugin is compatible |
```

#### Create your own GraphQL Type object

> In some cases, usually when an object is using some Generic Relationship, the default GraphQL `Type` object generated by the `extras_features` decorator is not working at the developer intends, and it is preferable to provide custom GraphQL types. A GraphQL `Type` object can be created and registered to the GraphQL interface by defining the type in the `graphql_types` variables in `graphql/types.py` file within the plugin. The object must inherit from `DjangoObjectType` and must follow the [standard defined by graphene-django](https://docs.graphene-python.org/projects/django/en/latest/queries/).

Reviewer: ```suggestion
In some cases, usually when an object is using some Generic Relationship, the default GraphQL `Type` object generated by the `extras_features` decorator may not work as the developer intends, and it will be preferable to provide custom GraphQL types. A GraphQL `Type` object can be created and registered to the GraphQL interface by defining the type in the `graphql_types` variables in `graphql/types.py` file within the plugin. The object must inherit from `DjangoObjectType` and must follow the [standard defined by graphene-django](https://docs.graphene-python.org/projects/django/en/latest/queries/).
```
